### PR TITLE
add custom string event

### PR DESCRIPTION
**Pull request by seedotlee**
You can set `customString` to change the color and click the string

### CODE REVIEW

**Comment by schickling** [comment]
Hi @seedotlee! Thank you for your pull request. I like the idea of having a way to customize the label configuration. However I think it would make more sense to implement this as a regex rule and not just a string comparison.

Please a lot try to be consistent about variable naming. For example `.CUSTOM` should be `.Custom`

What do you think?

**Comment by seedotlee** [comment]
OK, I will continue to improve

Next commit I will implement the regex rule.

**Comment by seedotlee** [comment]
Now ,I'm finished

**Comment by schickling** [comment]
Thanks @seedotlee. LGTM

@kaandedeoglu could you as a seasoned contributor also have a look at this? Thanks!

**Comment by kaandedeoglu** [comment]
@schickling @seedotlee I just had a look to the PR and here are my thoughts!

- Regex support is a fantastic addition - it gives the power user great flexibility, however it kind of feels incomplete if we are limited by just 1 regex string. I think we should have an array of search regexes - search priority going from earliest to latest. 

With how things are structured right now, maybe the regex handle blocks should have a signature of type `(String, String) -> Void` (instead of `String -> Void`) to inform the user which regex was responsible for the match

- I made some refactoring and improvements. I'll write them below:

<br>

#### `ActiveType.swift`
`RegexHelper` and the custom operator `=~` is not really needed. I think we should always think twice before adding a custom operator, it dirties the namespace and in this case it's not very helpful. We can just replace 

```swift
private struct RegexHelper {
    let regex: NSRegularExpression
    
    init(_ pattern: String) throws {
        try regex = NSRegularExpression(pattern: pattern,
            options: .CaseInsensitive)
    }
    
    func match(input: String) -> Array<NSTextCheckingResult> {
        let matches = regex.matchesInString(input,
            options: [],
            range: NSMakeRange(0, input.characters.count))
        return matches
    }
}

infix operator =~ {
    associativity none
    precedence 130
}

private func =~(lhs:String, rhs:String) -> Array<NSTextCheckingResult>? {
    do {
        return try RegexHelper(rhs).match(lhs)
    } catch _ {
        return nil
    }
}
```

with this 

```swift
private func regexMatches(regexString: String, searchString: String) -> Array<NSTextCheckingResult>? {
    guard let regex = try? NSRegularExpression(pattern: regexString, options: .CaseInsensitive) else { return nil }
    return regex.matchesInString(searchString, options: [], range: NSMakeRange(0, searchString.characters.count))
}
```

 this may not look as cool :) but is more readable imho


-------
You should use optional binding (`if let`) and optional chaining rather `!`. It makes expressions that are much nicer to read. 

```swift
if (regex != nil && (word=~regex!)?.count > 0) {
    return .Regex(word)
}
```

is more readable as (after the changes above)

```swift
if let regex = regex where regexMatches(regex, searchString: word)?.isEmpty == false {
    return .Regex(word)
}
```

<br>
#### `ActiveLabel.swift`
Something similar is also here:

```swift
let element:ActiveElement
        
if regexString != nil {
    element = activeElement(word, regex: regexString!)
} else {
    element = activeElement(word)
}
````

since the function takes an optional regex string and checks for presence - we can just replace the whole thing with

```swift
let element:ActiveElement = activeElement(word, regex: regexString)
```

**Comment by seedotlee** [comment]
@kaandedeoglu Thanks for your suggest that very useful for me and  it's  great code

I will change the code by your suggest.

**Comment by schickling** [comment]
Awesome @kaandedeoglu!

That's what I love about open source!

File: ActiveLabel/ActiveLabel.swift [hunk 1]
```diff
@@ -255,7 +284,14 @@ import UIKit
         var searchRange = NSMakeRange(0, textLength)
         
         for word in textString.componentsSeparatedByCharactersInSet(NSCharacterSet.whitespaceCharacterSet()) {
-            let element = activeElement(word)
+            
+            let element:ActiveElement
+            
+            if regexString != nil {
```
Comment (ajfigueroa) [inline ActiveLabel/ActiveLabel.swift]: Alternatively, you can check that an optional is not `nil` by using `if let` as follows:
```
if let regexString = regexString {
   // The unwrapped value is not nil
   // You can use regexString here with no need to force unwrap like regexString!
} else {
  // The unwrapped value is nil
}
```
This eliminates the need to force unwrap the regexString value

**Comment by kaandedeoglu** [comment]
@seedotlee what is the state with this PR? Are you still willing to make the discussed changes?

**Comment by seedotlee** [comment]
@kaandedeoglu I did finished it , but the code changed too much for my  business.

I will clear up the code and try to PR it.

**Comment by schickling** [comment]
Great. Thanks for the update @seedotlee @kaandedeoglu :+1:

**Comment by seedotlee** [comment]
I have PR the code please checked it.

 I'm sorry to I don't write the test case because I changed too much code what maybe you will not let it pass.

If you agreen the code I will add the test case.

@schickling @kaandedeoglu

**Comment by RHKliffer** [comment]
Hi @schickling 
When is this PR planned to be merged? I'd love to use this functionality

**Comment by Faisal-AlNeela** [comment]
Hi @seedotlee 
Why you didn't add the "add custom string" to the master project so we can update from cocoa pods??!

**Comment by dsinkovskiy** [comment]
please check my pull request #78. this problem was solved by adding public func addActiveElement(range: NSRange, type: ActiveType) method.

**Pull request closed without merging**